File: src/main/java/dragonbook/inter/stmt/Set.java
package dragonbook.inter.stmt;

import dragonbook.inter.expr.Expr;
import dragonbook.inter.expr.Id;
import dragonbook.symbols.Type;

public class Set extends Stmt {
    public Id id;
    public Expr expr;

    public Set(Id i, Expr x) {
        id = i;
        expr = x;
        if (check(id.type, expr.type) == null) error("type error");
    }

    public Type check(Type p1, Type p2) {
        if (Type.numeric(p1) && Type.numeric(p2)) return p2;
        else if (p1 == Type.BOOL && p2 == Type.BOOL) return p2;
        else return null;
    }
}

File: src/main/kotlin/dragonbook/inter/stmt/Set.kt
package dragonbook.inter.stmt

import dragonbook.error.ParseError
import dragonbook.inter.expr.Expr
import dragonbook.inter.expr.Id
import dragonbook.symbols.Type

data class Set(val id: Id, val expr: Expr) : Stmt {

    init {
        validate()
    }

    private fun validate() {
        if (id.type.isNumeric() && expr.type.isNumeric()) expr.type
        else if (id.type == Type.BOOL && expr.type == Type.BOOL) expr.type
        else throw ParseError(node = this)
    }

    override fun toString(): String = "$id = $expr"
}